### PR TITLE
Price feeds patch

### PR DESCRIPTION
**Pull request by VladyslavPerederWoof**
**Price Feeds Patch**

- **Zero collateral factor skip logic** - added logic to skip price feed processing for assets with zero collateral factors, preventing a broken/reverted price feed on one asset from paralyzing the entire protocol.

- **`AssetList.sol` validation update** - relaxed collateral factor validation to allow both `borrowCollateralFactor` and `liquidateCollateralFactor` to be set to `0` simultaneously (de-listed asset), while still enforcing ordering and range checks when both are non-zero.

- **`isBorrowCollateralized` / `isLiquidatable` resilience** - updated collateral iteration to skip assets with zero liquidation collateral factor, so a stale/reverting price feed on a de-listed asset doesn't block borrow or liquidation checks.

- **`quoteCollateral` fix** - updated discount calculation to use a local `liquidationFactor` variable and handle assets with zero CF correctly.

- **Whale addresses updated** - added new WBTC, ezETH, and weETH whale addresses to the test `WHALES` list for mainnet fork scenarios.

- **Scenarios & tests** - added and updated liquidation, withdraw, and quote collateral scenarios to cover zero-CF assets and extended asset list; added tests for price feed reverts.

### CODE REVIEW

File: contracts/AssetList.sol [hunk 1]
```diff
@@ -138,7 +138,7 @@ contract AssetList {
         if (IERC20NonStandard(asset).decimals() != decimals_) revert CometMainInterface.BadDecimals();
 
         // Ensure collateral factors are within range
-        if (assetConfig.borrowCollateralFactor >= assetConfig.liquidateCollateralFactor) revert CometMainInterface.BorrowCFTooLarge();
+        if (assetConfig.borrowCollateralFactor > MAX_COLLATERAL_FACTOR) revert CometMainInterface.BorrowCFTooLarge();
```
Comment (pavlo-h) [inline contracts/AssetList.sol]: the condition checked that borrowF < liquidateF
now condition is changed to borrowF <= 1, so now it will be possible to set borrow factor higher than the liquidate factor, which opens the way for misconfiguration

revert the condition back

File: contracts/CometWithExtendedAssetList.sol [hunk 2]
```diff
@@ -387,21 +387,31 @@ contract CometWithExtendedAssetList is CometMainInterface {
...
+        for (uint8 i; i < numAssets; ) {
             if (isInAsset(assetsIn, i, _reserved)) {
                 if (liquidity >= 0) {
                     return true;
                 }
 
-                AssetInfo memory asset = getAssetInfo(i);
-                uint newAmount = mulPrice(
+                asset = getAssetInfo(i);
+                borrowCollateralFactor = asset.borrowCollateralFactor;
```
Comment (pavlo-h) [inline contracts/CometWithExtendedAssetList.sol]: why do you need extra variable? you already have asset.borrowCollateralFactor

File: contracts/CometWithExtendedAssetList.sol [hunk 3]
```diff
@@ -430,21 +440,31 @@ contract CometWithExtendedAssetList is CometMainInterface {
...
+        for (uint8 i; i < numAssets; ) {
             if (isInAsset(assetsIn, i, _reserved)) {
                 if (liquidity >= 0) {
                     return false;
                 }
 
-                AssetInfo memory asset = getAssetInfo(i);
-                uint newAmount = mulPrice(
+                asset = getAssetInfo(i);
+                liquidateCollateralFactor = asset.liquidateCollateralFactor;
```
Comment (pavlo-h) [inline contracts/CometWithExtendedAssetList.sol]: why do you need extra variable? you already have asset.liquidateCOllateralFactor

File: contracts/CometWithExtendedAssetList.sol [hunk 4]
```diff
@@ -1161,20 +1181,31 @@ contract CometWithExtendedAssetList is CometMainInterface {
...
+        uint256 value;
+        uint64 liquidationFactor;
+        for (uint8 i; i < numAssets; ) {
             if (isInAsset(assetsIn, i, _reserved)) {
-                AssetInfo memory assetInfo = getAssetInfo(i);
-                address asset = assetInfo.asset;
-                uint128 seizeAmount = userCollateral[account][asset].balance;
+                assetInfo = getAssetInfo(i);
+
+                liquidationFactor = assetInfo.liquidationFactor;
```
Comment (pavlo-h) [inline contracts/CometWithExtendedAssetList.sol]: why do you need an extra variable?

File: contracts/CometWithExtendedAssetList.sol [hunk 5]
```diff
@@ -1249,15 +1280,21 @@ contract CometWithExtendedAssetList is CometMainInterface {
     function quoteCollateral(address asset, uint baseAmount) override public view returns (uint) {
         AssetInfo memory assetInfo = getAssetInfoByAddress(asset);
         uint256 assetPrice = getPrice(assetInfo.priceFeed);
-        // Store front discount is derived from the collateral asset's liquidationFactor and storeFrontPriceFactor
-        // discount = storeFrontPriceFactor * (1e18 - liquidationFactor)
-        uint256 discountFactor = mulFactor(storeFrontPriceFactor, FACTOR_SCALE - assetInfo.liquidationFactor);
-        uint256 assetPriceDiscounted = mulFactor(assetPrice, FACTOR_SCALE - discountFactor);
+        uint64 liquidationFactor = assetInfo.liquidationFactor;
```
Comment (pavlo-h) [inline contracts/CometWithExtendedAssetList.sol]: why do you need an extra variable here

File: contracts/CometWithExtendedAssetList.sol [hunk 6]
```diff
@@ -1249,15 +1280,21 @@ contract CometWithExtendedAssetList is CometMainInterface {
...
+
+        // If liquidation factor is not zero, calculate the discount
+        if (liquidationFactor != 0) {
+            // Store front discount is derived from the collateral asset's liquidationFactor and storeFrontPriceFactor
+            // discount = storeFrontPriceFactor * (1e18 - liquidationFactor)
+            uint256 discountFactor = mulFactor(storeFrontPriceFactor, FACTOR_SCALE - liquidationFactor);
+            assetPrice = mulFactor(assetPrice, FACTOR_SCALE - discountFactor);
```
Comment (pavlo-h) [inline contracts/CometWithExtendedAssetList.sol]: why renamed from assetPriceDiscounted?

File: contracts/CometWithExtendedAssetList.sol [hunk 7]
```diff
@@ -1249,15 +1280,21 @@ contract CometWithExtendedAssetList is CometMainInterface {
     function quoteCollateral(address asset, uint baseAmount) override public view returns (uint) {
         AssetInfo memory assetInfo = getAssetInfoByAddress(asset);
         uint256 assetPrice = getPrice(assetInfo.priceFeed);
-        // Store front discount is derived from the collateral asset's liquidationFactor and storeFrontPriceFactor
-        // discount = storeFrontPriceFactor * (1e18 - liquidationFactor)
-        uint256 discountFactor = mulFactor(storeFrontPriceFactor, FACTOR_SCALE - assetInfo.liquidationFactor);
-        uint256 assetPriceDiscounted = mulFactor(assetPrice, FACTOR_SCALE - discountFactor);
+        uint64 liquidationFactor = assetInfo.liquidationFactor;
+
+        // If liquidation factor is not zero, calculate the discount
```
Comment (pavlo-h) [inline contracts/CometWithExtendedAssetList.sol]: and what if zero? what will happen? countereffect is not commented, so the goal of if is unclear

File: deployments/hardhat/dai/deploy.ts [hunk 8]
```diff
@@ -63,6 +63,7 @@ export default async function deploy(deploymentManager: DeploymentManager, deplo
   const deployed = await deployComet(deploymentManager, deploySpec, {
     baseTokenPriceFeed: daiPriceFeed.address,
     assetConfigs: [assetConfig0, assetConfig1],
+    withMockAssetListFactory: true,
```
Comment (pavlo-h) [inline deployments/hardhat/dai/deploy.ts]: ?

File: contracts/CometWithExtendedAssetList.sol [hunk 9]
```diff
@@ -430,21 +440,31 @@ contract CometWithExtendedAssetList is CometMainInterface {
...
                 if (liquidity >= 0) {
                     return false;
                 }
 
-                AssetInfo memory asset = getAssetInfo(i);
-                uint newAmount = mulPrice(
+                asset = getAssetInfo(i);
+                liquidateCollateralFactor = asset.liquidateCollateralFactor;
+                
+                if (liquidateCollateralFactor == 0) {
```
Comment (pavlo-h) [inline contracts/CometWithExtendedAssetList.sol]: add a comment why we need this condition
do not hesitate to explain the code in comments if necessary

File: contracts/CometWithExtendedAssetList.sol [hunk 10]
```diff
@@ -387,21 +387,31 @@ contract CometWithExtendedAssetList is CometMainInterface {
...
                 if (liquidity >= 0) {
                     return true;
                 }
 
-                AssetInfo memory asset = getAssetInfo(i);
-                uint newAmount = mulPrice(
+                asset = getAssetInfo(i);
+                borrowCollateralFactor = asset.borrowCollateralFactor;
+
+                if (borrowCollateralFactor == 0) {
```
Comment (pavlo-h) [inline contracts/CometWithExtendedAssetList.sol]: add a comment why we need this condition
do not hesitate to explain the code in comments if necessary

File: contracts/AssetList.sol [hunk 11]
```diff
@@ -148,9 +148,6 @@ contract AssetList {
             uint16 liquidateCollateralFactor = uint16(assetConfig.liquidateCollateralFactor / descale);
             uint16 liquidationFactor = uint16(assetConfig.liquidationFactor / descale);
 
-            // Be nice and check descaled values are still within range
-            if (borrowCollateralFactor >= liquidateCollateralFactor) revert CometMainInterface.BorrowCFTooLarge();
```
Comment (pavlo-h) [inline contracts/AssetList.sol:152]: why have you removed this condition?

File: contracts/CometWithExtendedAssetList.sol [hunk 12]
```diff
@@ -1161,20 +1181,31 @@ contract CometWithExtendedAssetList is CometMainInterface {
...
+        uint64 liquidationFactor;
+        for (uint8 i; i < numAssets; ) {
             if (isInAsset(assetsIn, i, _reserved)) {
-                AssetInfo memory assetInfo = getAssetInfo(i);
-                address asset = assetInfo.asset;
-                uint128 seizeAmount = userCollateral[account][asset].balance;
+                assetInfo = getAssetInfo(i);
+
+                liquidationFactor = assetInfo.liquidationFactor;
+                if (liquidationFactor == 0) {
```
Comment (pavlo-h) [inline contracts/CometWithExtendedAssetList.sol]: add a comment why we need this condition
do not hesitate to explain the code in comments if necessary

File: contracts/CometWithExtendedAssetList.sol [hunk 13]
```diff
@@ -1249,15 +1280,21 @@ contract CometWithExtendedAssetList is CometMainInterface {
...
         AssetInfo memory assetInfo = getAssetInfoByAddress(asset);
         uint256 assetPrice = getPrice(assetInfo.priceFeed);
-        // Store front discount is derived from the collateral asset's liquidationFactor and storeFrontPriceFactor
-        // discount = storeFrontPriceFactor * (1e18 - liquidationFactor)
-        uint256 discountFactor = mulFactor(storeFrontPriceFactor, FACTOR_SCALE - assetInfo.liquidationFactor);
-        uint256 assetPriceDiscounted = mulFactor(assetPrice, FACTOR_SCALE - discountFactor);
+        uint64 liquidationFactor = assetInfo.liquidationFactor;
+
+        // If liquidation factor is not zero, calculate the discount
+        if (liquidationFactor != 0) {
```
Comment (pavlo-h) [inline contracts/CometWithExtendedAssetList.sol]: comment that the only purpose of this if is to have market price when factor == 0

File: contracts/CometWithExtendedAssetList.sol (hunk 7 above)
Comment (pavlo-h) [reply contracts/CometWithExtendedAssetList.sol]: add a comment why we need this condition
do not hesitate to explain the code in comments if necessary

File: contracts/CometWithExtendedAssetList.sol [hunk 14]
```diff
@@ -1249,15 +1280,21 @@ contract CometWithExtendedAssetList is CometMainInterface {
     function quoteCollateral(address asset, uint baseAmount) override public view returns (uint) {
         AssetInfo memory assetInfo = getAssetInfoByAddress(asset);
         uint256 assetPrice = getPrice(assetInfo.priceFeed);
```
Comment (pavlo-h) [inline contracts/CometWithExtendedAssetList.sol]: getPrice is here unattended so quote will still revert when pricefeed is disabled - comment it somewhere in scenarios

File: scenario/LiquidationScenario.ts [hunk 15]
```diff
@@ -297,4 +297,214 @@ scenario.skip(
...
+
+      expect(await comet.isLiquidatable(albert.address)).to.be.true;
+
+      // Save balances before absorb
+      const userCollateralBefore = (await comet.userCollateral(albert.address, asset)).balance;
+      const totalsBefore = (await comet.totalsCollateral(asset)).totalSupplyAsset;
+
+      await betty.absorb({ absorber: betty.address, accounts: [albert.address] });
+
+      expect((await comet.userCollateral(albert.address, asset)).balance).to.equal(userCollateralBefore);
```
Comment (pavlo-h) [inline scenario/LiquidationScenario.ts:505]: you test absorbtion - that the debt will be absorbed for a collateral with liquidation collateral factor = 0
but there are no check on that. You check total balances, but not the user's debt. How can we be sure that debt is definitely absrobed?

**Review by pavlo-h — CHANGES_REQUESTED** [review]

File: contracts/AssetList.sol (hunk 1 above)
Comment (pavlo-h) [reply contracts/AssetList.sol]: applied edgecases (as with liquidate factor == 0) should not break the rest of the logic
currently the logic that liquidation factor always higher than borrow factor breaks after the update

Comment (pavlo-h) [reply contracts/AssetList.sol]: require(assetConfig.borrowCollateralFactor < assetConfig.liquidateCollateralFactor || assetConfig.liquidateCollateralFactor == 0) or smth like that should work

but also - important question - if both borrow and liquidation factor will be set to 0 simultaneously, or only one of them in time? this answer should be documented and implemented in appropriate conditions

File: contracts/CometWithExtendedAssetList.sol (hunk 2 above)
Comment (VladyslavPerederWoof) [reply contracts/CometWithExtendedAssetList.sol]: Gas estimation checks: 

CF != 0 
With separate variable: isBorrowCollateralized gas estimate: 79034 units 
Directly: isBorrowCollateralized gas estimate: 79123 units  

CF = 0
With separate variable: isBorrowCollateralized gas estimate: 61917 units 
Directly: isBorrowCollateralized gas estimate: 61917 units

Comment (VladyslavPerederWoof) [reply contracts/CometWithExtendedAssetList.sol]: The check was done with only 1 collateral

Comment (pavlo-h) [reply contracts/CometWithExtendedAssetList.sol]: I guess compiler will optimize it anyway, let it be as microoptimization

Comment (VladyslavPerederWoof) [reply contracts/CometWithExtendedAssetList.sol]: Separate variable removed 👌

File: contracts/CometWithExtendedAssetList.sol (hunk 4 above)
Comment (VladyslavPerederWoof) [reply contracts/CometWithExtendedAssetList.sol]: Separate variable removed 👌

File: contracts/CometWithExtendedAssetList.sol (hunk 2 above)
Comment (pavlo-h) [reply contracts/CometWithExtendedAssetList.sol]: though pay attention - here you have 2 accesses to that structure field, so you optimize it into 1 access
in some other places you may have 1 access to asset.field - and so you just add new variable upon it with no optimizations.

File: contracts/AssetList.sol (hunk 1 above)
Comment (VladyslavPerederWoof) [reply contracts/AssetList.sol]: What if we make a check for factors that should be in a range only in case:

```
// Ensure collateral factors are within range
if (assetConfig.borrowCollateralFactor != 0 && assetConfig.liquidateCollateralFactor != 0) {
            if (assetConfig.borrowCollateralFactor > assetConfig.liquidateCollateralFactor) revert CometMainInterface.BorrowCFTooLarge();
            if (assetConfig.liquidateCollateralFactor > MAX_COLLATERAL_FACTOR) revert CometMainInterface.LiquidateCFTooLarge();
}
```

Comment (VladyslavPerederWoof) [reply contracts/AssetList.sol]: Same will do when factors are descaled later in code

File: contracts/CometWithExtendedAssetList.sol (hunk 2 above)
Comment (VladyslavPerederWoof) [reply contracts/CometWithExtendedAssetList.sol]: Totally agree, in case of 1 access to asset.field will have no sense to move it to the separate variable

File: deployments/hardhat/dai/deploy.ts (hunk 8 above)
Comment (VladyslavPerederWoof) [reply deployments/hardhat/dai/deploy.ts]: Was added during progress. I forgot to remove this. Fixed

File: contracts/CometWithExtendedAssetList.sol (hunk 6 above)
Comment (VladyslavPerederWoof) [reply contracts/CometWithExtendedAssetList.sol]: Now we have to 2 possible scenarios: when price will be with discount and without
As we might have price without discount, I decided to change name to more flexible

File: contracts/CometWithExtendedAssetList.sol (hunk 14 above)
Comment (VladyslavPerederWoof) [reply contracts/CometWithExtendedAssetList.sol]: Comment directly in sc?

File: scenario/LiquidationScenario.ts (hunk 15 above)
Comment (VladyslavPerederWoof) [reply scenario/LiquidationScenario.ts:505]: I don't quite understand. When the liquidation factor is zero, we can absorb a user, but absorption will be skipped for that asset, as the factor is zero

File: contracts/CometWithExtendedAssetList.sol (hunk 7 above)
Comment (VladyslavPerederWoof) [reply contracts/CometWithExtendedAssetList.sol]: What about?
 ```
   // Only apply the store front discount for assets that participate in liquidation
        // (i.e. liquidationFactor > 0). Assets with liquidationFactor == 0 are non-liquidatable:
        // they are skipped during absorption (see absorbInternal) and therefore should not
        // receive a liquidation discount when purchased via buyCollateral.
        //
        // Additionally, if liquidationFactor == 0 the discount math would compute
        // discountFactor = storeFrontPriceFactor * (FACTOR_SCALE - 0) / FACTOR_SCALE
        //                = storeFrontPriceFactor,
        // and when storeFrontPriceFactor == FACTOR_SCALE (100%) that yields
        // assetPrice = assetPrice * (FACTOR_SCALE - FACTOR_SCALE) / FACTOR_SCALE = 0,
        // which would cause a division-by-zero revert in the return statement below.
        //
        // By skipping the discount, the protocol sells such collateral at the fair oracle
        // price — no liquidation incentive is needed for non-liquidatable assets.
```

File: contracts/CometWithExtendedAssetList.sol (hunk 14 above)
Comment (pavlo-h) [reply contracts/CometWithExtendedAssetList.sol]: you have patch to address revert for inadequate price feed - you wrap each price feed call into the condition of factor being 0. Here you still have direct call which will revert in case of broken price feed. We need it for a reason - as quote should always work.
So - put comments here in SC to show that it is left deliberately and that buyCollateral will revert for inadequate pricefeeds, but account can still be absorbed to avoid bad debt

and same goes into the scenario where quote is tested

File: contracts/CometWithExtendedAssetList.sol (hunk 6 above)
Comment (pavlo-h) [reply contracts/CometWithExtendedAssetList.sol]: so again - we are talking about edge case influencing main logic
leave the original name as it shows the intent

File: scenario/LiquidationScenario.ts (hunk 15 above)
Comment (pavlo-h) [reply scenario/LiquidationScenario.ts:505]: yes, you test that
but where is the check that debt is absorbed?

File: contracts/CometWithExtendedAssetList.sol (hunk 6 above)
Comment (VladyslavPerederWoof) [reply contracts/CometWithExtendedAssetList.sol]: 👌

File: scenario/LiquidationScenario.ts (hunk 15 above)
Comment (VladyslavPerederWoof) [reply scenario/LiquidationScenario.ts:505]: Oh, in case when 2 collaterals and 1 is with factor 0
And check that 1 will be absorbed and other one will be skipped
I got a point, will add it soon

File: contracts/AssetList.sol (hunk 1 above)
Comment (pavlo-h) [reply contracts/AssetList.sol]: plus

Comment (pavlo-h) [reply contracts/AssetList.sol]: and add a comment that for such de-listing we require both parameters to be 0

File: contracts/AssetList.sol [hunk 16]
```diff
@@ -137,9 +137,15 @@ contract AssetList {
...
-        if (assetConfig.liquidateCollateralFactor > MAX_COLLATERAL_FACTOR) revert CometMainInterface.LiquidateCFTooLarge();
+        // To de-list (disable) an asset as collateral, both borrowCollateralFactor and
+        // liquidateCollateralFactor must be set to 0. Setting only one to 0 is not
+        // supported — the validation below is skipped entirely when either factor is 0,
+        // so a partial zero configuration would bypass the range and ordering checks.
+        if (assetConfig.borrowCollateralFactor != 0 && assetConfig.liquidateCollateralFactor != 0) {
+            // Ensure collateral factors are within range
+            if (assetConfig.borrowCollateralFactor > assetConfig.liquidateCollateralFactor) revert CometMainInterface.BorrowCFTooLarge();
+            if (assetConfig.liquidateCollateralFactor > MAX_COLLATERAL_FACTOR) revert CometMainInterface.LiquidateCFTooLarge();
+        }
```
Comment (beiraoCertora) [inline contracts/AssetList.sol:151]: I suppose it would make sense to revert if `borrowCollateralFactor` XOR  `liquidateCollateralFactor` is 0.

Because it will mean the asset is missconfigured.


```solidity
if (assetConfig.borrowCollateralFactor == 0 && assetConfig.liquidateCollateralFactor == 0) {
    // De-listed: skip validation
} else if (assetConfig.borrowCollateralFactor != 0 && assetConfig.liquidateCollateralFactor != 0) {
    // Both non-zero: validate ordering
    if (assetConfig.borrowCollateralFactor > assetConfig.liquidateCollateralFactor) revert CometMainInterface.BorrowCFTooLarge();
    if (assetConfig.liquidateCollateralFactor > MAX_COLLATERAL_FACTOR) revert CometMainInterface.LiquidateCFTooLarge();
} else {
    // One zero, one non-zero: reject
    revert CometMainInterface.BadAssetConfiguration();
}
```
